File: exercises/exercism/ruby/hamming.rb
# Hamming from https://exercism.io

class Hamming
  def self.compute(x, y)
    self.new(x, y).compute
  end

  def initialize(x, y)
    @x, @y = x, y
  end

  def compute
    raise ArgumentError if x.length != y.length

    x.chars.zip(y.chars).count { |xx, yy| xx != yy }
  end

  private

    attr_reader :x, :y
end

# Array#zip
# https://docs.ruby-lang.org/ja/2.6.0/method/Array/i/zip.html
# Enumerable#count
# https://docs.ruby-lang.org/ja/2.6.0/method/Enumerable/i/count.html
# ブロックを渡すことができる

File: exercises/exercism/c/2020/hamming.c
/* Hamming from https://exercism.io */

#include "hamming.h"
#include <string.h>

int compute(const char *lhs, const char *rhs)
{
  int is_null = (lhs == NULL) || (rhs == NULL);

  if (is_null || (int)strlen(lhs) != (int)strlen(rhs)) {
    return -1;
  }

  int count = 0;

  for (int i = 0; i < (int)strlen(lhs); i++) {
    if (lhs[i] != rhs[i]) {
      count += 1;
    }
  }
  return count;
}
